CASMCMS-9513: Known issue IMS Image Customization Job Status Stuck at `waiting_on_user` (#6198)
Cray-HPE/docs-csm: troubleshooting/known_issues/ims_image_customization_job_status_stuck_at_waiting_on_user.md

# IMS Image Customization Job Status Stuck at "waiting_on_user"

## Issue description

> IMS image customization job on remote node can get stuck in the "waiting_on_user" state indefinitely.

Reviewer: ```suggestion
An IMS image customization job on a remote node can get stuck in the `waiting_on_user` state indefinitely.
```

> This can occur if during `image customization` remote node gets rebooted, crashes or IMS JOB container is killed/stopped.

Reviewer: ```suggestion
This can occur during `image customization` if any of the following things happen:

- The remote node reboots or crashes
- The IMS job container is killed or stopped.
```

## Error identification

The issue can be identified by noticing the failure to ssh into `sshd` container of IMS image customization job. here are

> the steps to identify the issue:

Reviewer: ```suggestion
A symptom of this problem is a failure when attempting to SSH into the `sshd` container of the IMS image
customization job. Use the following procedure to detect the issue.
```

> 1. (`ncn-mw#`) Get the details of the image customization job:
> 
>    ```bash
>    IMS_JOB_ID=<Job ID>
>    cray ims jobs describe $IMS_JOB_ID --format json
>    ```

Reviewer: ````suggestion
1. (`ncn-mw#`) Get the details of the image customization job,

   > In the following command, substitute the actual IMS job ID being checked.

   ```bash
   cray ims jobs describe <IMS_JOB_ID> --format json
   ```
````

>   Job details will show the job status as "waiting_on_user". here is sample output:
>   
>   ```text

Reviewer: ````suggestion
  Example output:
  
  ```json
````

  "kubernetes_configmap": "cray-ims-10b93eb5-2926-4521-8cd8-bcb9ab92f989-configmap",
  "kubernetes_job": "cray-ims-10b93eb5-2926-4521-8cd8-bcb9ab92f989-customize",
  "kubernetes_namespace": "ims",
  "kubernetes_pvc": "cray-ims-10b93eb5-2926-4521-8cd8-bcb9ab92f989-job-claim",
  "kubernetes_secret": "cray-ims-10b93eb5-2926-4521-8cd8-bcb9ab92f989-signing-keys",
  "kubernetes_service": "cray-ims-10b93eb5-2926-4521-8cd8-bcb9ab92f989-service",
  "public_key_id": "7d560617-91e1-4075-b8ab-891a3285b783",
  "remote_build_node": "x3000c0s33b1n0",
  "require_dkms": true,
  "resultant_image_id": null,
  "ssh_containers": [
    {
      "connection_info": {
        "cluster.local": {
          "host": "cray-ims-10b93eb5-2926-4521-8cd8-bcb9ab92f989-service.ims.svc.cluster.local",
          "port": 22
        },
        "customer_access": {
          "host": "10b93eb5-2926-4521-8cd8-bcb9ab92f989.ims.cmn.fanta.hpc.amslabs.hpecorp.net",
          "port": 22
        }
      },
      "jail": false,
      "name": "sat-0fda73ef-4ab0-46d2-9e36-9aad6406bb50",
      "status": "pending"[README.md](../README.md)
    }
  ],
  "status": "waiting_on_user"
  }
  ```

> 

Reviewer: ```suggestion

1. Confirm that the job details show the job status as `waiting_on_user`.

   If that is not the case, then the procedure documented here is not applicable.

```

> 1. (`ncn-mw#`) Attempt to ssh into the `sshd` container of the job and notice the connection failure:
> 
>     ```bash
>    IMS_SSH_HOST=<ssh_containers[0].connection_info.customer_access.host from above output>
>    IMS_SSH_PORT=<ssh_containers[0].connection_info.customer_access.port from above output>
>    ssh -p $IMS_SSH_PORT root@$IMS_SSH_HOST
>    ```

Reviewer: ````suggestion
1. (`ncn-mw#`) Attempt to SSH into the `sshd` container of the job.

   > Perform the following substitutions in the command:
   >
   > - Replace `<IMS_SSH_HOST>` with the value of the
   >   `ssh_containers[0].connection_info.customer_access.host` field in the job details.
   > - Replace `<IMS_SSH_POST>` with the value of the
   >   `ssh_containers[0].connection_info.customer_access.port` field in the job details.

   ```bash
   ssh -p <IMS_SSH_PORT> root@<IMS_SSH_HOST>
   ```

1. Confirm that the SSH attempt resulted in a connection failure.

   If that is not the case, then the procedure documented here is not applicable.

````

> 1. (`ncn-mw#`) connect to the remote node where the job is running:

Reviewer: ```suggestion
1. (`ncn-mw#`) Connect to the remote node where the job is running.
```

   ```bash
   ssh <remote_build_node xname>
   ```

> if the remote node is not reachable then skip the next step and go to the [resolution](#resolution) section.

Reviewer: ````suggestion
   > If the remote node is not reachable, then skip ahead to the [Resolution](#resolution) section.

   ```bash
   ssh <remote_build_node xname>
   ```
````

> 1. (`ncn-mw#`) Check if `ims` job container exists on the remote node:
> 
>    ```bash
>    IMS_JOB_ID=<Job ID>
>    podman ps | grep $IMS_JOB_ID
>    ```
> 
>    You will notice that either `ims` job container does not exist or is in `exited` state.

Reviewer: ````suggestion
1. (`cn#`) Check if IMS job container exists on the remote node.

   > In the following command, replace `<IMS_JOB_ID>` with the actual IMS job ID.

   ```bash
   podman ps | grep <IMS_JOB_ID>
   ```

1, If the IMS job container either does not exist or is in an `exited` state, then proceed to [Resolution](#resolution),

   If that is not the case, then the procedure documented here is not applicable.
````

## Resolution

> In this case the `ims` job needs to be removed. Follow these steps:
> 
> 1. (`ncn-mw#`) Remove the `ims` image customization job using `cray` CLI:
> 
>    ```bash
>    IMS_JOB_ID=<Job ID>
>    cray ims jobs delete $IMS_JOB_ID
>    ```
> 
> once the job is deleted a new image customization job can be created using following
> [IMS Image Customization](../../operations/image_management/Customize_an_Image_Root_Using_IMS.md)

Reviewer: ````suggestion
(`ncn-mw#`) In order to resolve the problem, delete the IMS job.

> In the following command, replace `<IMS_JOB_ID>` with the actual IMS job ID.

 ```bash
cray ims jobs delete <IMS_JOB_ID>
```

After the job is deleted, a new image customization job can be created. See
[Customize an Image Root Using IMS](../../operations/image_management/Customize_an_Image_Root_Using_IMS.md).
````

# IMS Image Customization Job Status Stuck at `waiting_on_user`

## Issue description

An IMS image customization job on a remote node can get stuck in the `waiting_on_user` state indefinitely.
This can occur during `image customization` if any of the following things happen:

- The remote node reboots or crashes

> - The IMS job container is killed or stopped.

Reviewer 2: 'The IMS job container on the remote node is killed or stopped.'

Reviewer: ```suggestion
- The IMS job container on the remote build node is killed or stopped.
```

  "kernel_parameters_file_name": "kernel-parameters",
  "kubernetes_configmap": "cray-ims-10b93eb5-2926-4521-8cd8-bcb9ab92f989-configmap",
  "kubernetes_job": "cray-ims-10b93eb5-2926-4521-8cd8-bcb9ab92f989-customize",
  "kubernetes_namespace": "ims",
  "kubernetes_pvc": "cray-ims-10b93eb5-2926-4521-8cd8-bcb9ab92f989-job-claim",
  "kubernetes_secret": "cray-ims-10b93eb5-2926-4521-8cd8-bcb9ab92f989-signing-keys",
  "kubernetes_service": "cray-ims-10b93eb5-2926-4521-8cd8-bcb9ab92f989-service",
  "public_key_id": "7d560617-91e1-4075-b8ab-891a3285b783",
  "remote_build_node": "x3000c0s33b1n0",
  "require_dkms": true,
  "resultant_image_id": null,
  "ssh_containers": [
    {
      "connection_info": {
        "cluster.local": {
          "host": "cray-ims-10b93eb5-2926-4521-8cd8-bcb9ab92f989-service.ims.svc.cluster.local",
          "port": 22
        },
        "customer_access": {
          "host": "10b93eb5-2926-4521-8cd8-bcb9ab92f989.ims.cmn.fanta.hpc.amslabs.hpecorp.net",
          "port": 22
        }
      },
      "jail": false,
      "name": "sat-0fda73ef-4ab0-46d2-9e36-9aad6406bb50",
      "status": "pending"
    }
  ],
  "status": "waiting_on_user"
  }
  ```

> 1. Confirm that the job details show the job status as `waiting_on_user`.

Reviewer 2: " `waiting_on_user` and running on a remote node."

Author: done

   If that is not the case, then the procedure documented here is not applicable.

1. (`ncn-mw#`) Attempt to `ssh` into the `sshd` container of the job and notice the connection failure:

    ```bash
   IMS_SSH_HOST=<ssh_containers[0].connection_info.customer_access.host from above output>
   IMS_SSH_PORT=<ssh_containers[0].connection_info.customer_access.port from above output>
   ssh -p $IMS_SSH_PORT root@$IMS_SSH_HOST
   ```

1. (`ncn-mw#`) Connect to the remote node where the job is running.

>    > If the remote node is not reachable, then skip ahead to the [Resolution](#resolution) section.

Reviewer: ```suggestion
   > - The name of the remote build node can be found in the `remote_build_node` field of the job details.
   > - If the remote node is not reachable, then skip ahead to the [Resolution](#resolution) section.
```

   ```bash
   ssh <remote_build_node xname>
   ```

> 1. (`ncn-mw#`) Check if `ims` job container exists on the remote node:

Reviewer 2: This needs to be run on the remote node, not the ncn's.

Author: Done

  "public_key_id": "7d560617-91e1-4075-b8ab-891a3285b783",
  "remote_build_node": "x3000c0s33b1n0",
  "require_dkms": true,
  "resultant_image_id": null,
  "ssh_containers": [
    {
      "connection_info": {
        "cluster.local": {
          "host": "cray-ims-10b93eb5-2926-4521-8cd8-bcb9ab92f989-service.ims.svc.cluster.local",
          "port": 22
        },
        "customer_access": {
          "host": "10b93eb5-2926-4521-8cd8-bcb9ab92f989.ims.cmn.fanta.hpc.amslabs.hpecorp.net",
          "port": 22
        }
      },
      "jail": false,
      "name": "sat-0fda73ef-4ab0-46d2-9e36-9aad6406bb50",
      "status": "pending"
    }
  ],
  "status": "waiting_on_user"
  }
  ```

1. Confirm that the job details show the job status as `waiting_on_user`.

   If that is not the case, then the procedure documented here is not applicable.

1. (`ncn-mw#`) Attempt to `ssh` into the `sshd` container of the job and notice the connection failure:

    ```bash
   IMS_SSH_HOST=<ssh_containers[0].connection_info.customer_access.host from above output>
   IMS_SSH_PORT=<ssh_containers[0].connection_info.customer_access.port from above output>

>    ssh -p $IMS_SSH_PORT root@$IMS_SSH_HOST

Reviewer: Don't we have to make sure to use the right SSH key, or it could get rejected on that basis? Or will that generate a failure but a different one?

Author: ssh public key is supllied as part of the customize job command. The customize image doc does not mention use ssh key while ssh into sshd container
https://github.com/Cray-HPE/docs-csm/blob/release/1.7/operations/image_management/Customize_an_Image_Root_Using_IMS.md

Reviewer: I think in that procedure, it says to make sure that the key you use is the one associated with the current user (the one creating the job). So if you are still logged in as that same user with the same credentials, then it would presumably just work, since it's using your public key.

But if the person looking into the hang isn't the same user, then I assume the SSH would fail, unless they happen to share a public SSH key. You may want to check with @dlaine-hpe to be sure, though -- that's just how I thought it worked, but I am not an IMS expert.

Author: Added a note after discussing with David.

Reviewer: I don't see the note in the PR. Where is it?

Reviewer: I see it now.

  "kubernetes_job": "cray-ims-10b93eb5-2926-4521-8cd8-bcb9ab92f989-customize",
  "kubernetes_namespace": "ims",
  "kubernetes_pvc": "cray-ims-10b93eb5-2926-4521-8cd8-bcb9ab92f989-job-claim",
  "kubernetes_secret": "cray-ims-10b93eb5-2926-4521-8cd8-bcb9ab92f989-signing-keys",
  "kubernetes_service": "cray-ims-10b93eb5-2926-4521-8cd8-bcb9ab92f989-service",
  "public_key_id": "7d560617-91e1-4075-b8ab-891a3285b783",
  "remote_build_node": "x3000c0s33b1n0",
  "require_dkms": true,
  "resultant_image_id": null,
  "ssh_containers": [
    {
      "connection_info": {
        "cluster.local": {
          "host": "cray-ims-10b93eb5-2926-4521-8cd8-bcb9ab92f989-service.ims.svc.cluster.local",
          "port": 22
        },
        "customer_access": {
          "host": "10b93eb5-2926-4521-8cd8-bcb9ab92f989.ims.cmn.fanta.hpc.amslabs.hpecorp.net",
          "port": 22
        }
      },
      "jail": false,
      "name": "sat-0fda73ef-4ab0-46d2-9e36-9aad6406bb50",
      "status": "pending"
    }
  ],
  "status": "waiting_on_user"
  }
  ```

1. Confirm that the job details show the job status as `waiting_on_user` and running on a remote node..

>    If that is not the case, then the procedure documented here is not applicable.

Reviewer: ```suggestion
1. Confirm that the job details fit the symptoms of this issue.

   - The job `status` field is `waiting_on_user`
   - The `remote_build_node` field is set, indicating the job is running on a remote build node.

   If either of these is not the case, then the procedure documented here is not applicable.
```

1. (`ncn-mw#`) Attempt to SSH into the `sshd` container of the job.

   Perform the following substitutions in the command:

   - Replace `<IMS_SSH_HOST>` with the value of the
     `ssh_containers[0].connection_info.customer_access.host` field in the job details.
   - Replace `<IMS_SSH_POST>` with the value of the

>      `ssh_containers[0].connection_info.customer_access.port` field in the job details.

Reviewer: ```suggestion
   > Perform the following substitutions in the command:
   > 
   > - Replace `<IMS_SSH_HOST>` with the value of the
   >    `ssh_containers[0].connection_info.customer_access.host` field in the job details.
   > - Replace `<IMS_SSH_PORT>` with the value of the
   >   `ssh_containers[0].connection_info.customer_access.port` field in the job details.
```

# IMS Image Customization Job Status Stuck at `waiting_on_user`

## Issue description

> An IMS image customization job on a remote node can get stuck in the `waiting_on_user` state indefinitely.
> This can occur during `image customization` if any of the following things happen:
> 
> - The remote node reboots or crashes

Reviewer: ```suggestion
An IMS image customization job on a remote build node can get stuck in the `waiting_on_user` state indefinitely.
This can occur during `image customization` if any of the following things happen:

- The remote build node reboots or crashes
```

  "kubernetes_service": "cray-ims-10b93eb5-2926-4521-8cd8-bcb9ab92f989-service",
  "public_key_id": "7d560617-91e1-4075-b8ab-891a3285b783",
  "remote_build_node": "x3000c0s33b1n0",
  "require_dkms": true,
  "resultant_image_id": null,
  "ssh_containers": [
    {
      "connection_info": {
        "cluster.local": {
          "host": "cray-ims-10b93eb5-2926-4521-8cd8-bcb9ab92f989-service.ims.svc.cluster.local",
          "port": 22
        },
        "customer_access": {
          "host": "10b93eb5-2926-4521-8cd8-bcb9ab92f989.ims.cmn.fanta.hpc.amslabs.hpecorp.net",
          "port": 22
        }
      },
      "jail": false,
      "name": "sat-0fda73ef-4ab0-46d2-9e36-9aad6406bb50",
      "status": "pending"
    }
  ],
  "status": "waiting_on_user"
  }
  ```

1. Confirm that the job details fit the symptoms of this issue.

   - The job `status` field is `waiting_on_user`
   - The `remote_build_node` field is set, indicating the job is running on a remote build node.

   If either of these is not the case, then the procedure documented here is not applicable.

1. (`ncn-mw#`) Attempt to SSH into the `sshd` container of the job.

   > Perform the following substitutions in the command:

>    > 

Reviewer: ```suggestion
   >
```

   > - Replace `<IMS_SSH_HOST>` with the value of the
   >    `ssh_containers[0].connection_info.customer_access.host` field in the job details.
   > - Replace `<IMS_SSH_PORT>` with the value of the
   >   `ssh_containers[0].connection_info.customer_access.port` field in the job details.

   ```bash
   ssh -p <IMS_SSH_PORT> root@<IMS_SSH_HOST>
   ```

1. Confirm that the SSH attempt resulted in a connection failure.
   If that is not the case, then the procedure documented here is not applicable.

1. (`ncn-mw#`) Connect to the remote node where the job is running.

   > - The name of the remote build node can be found in the `remote_build_node` field of the job details.
   > - If the remote node is not reachable, then skip ahead to the [Resolution](#resolution) section.

   ```bash
   ssh <remote_build_node xname>
   ```

1. (`cn#`) Check if IMS job container exists on the remote node.

   > In the following command, replace `<IMS_JOB_ID>` with the actual IMS job ID.

   ```bash
   podman ps | grep <IMS_JOB_ID>
   ```

> 1, If the IMS job container either does not exist or is in an `exited` state, then proceed to [Resolution](#resolution),

Reviewer: ```suggestion
1. If the IMS job container either does not exist or is in an `exited` state, then proceed to [Resolution](#resolution),
```

# IMS Image Customization Job Status Stuck at `waiting_on_user`

## Issue description

An IMS image customization job on a remote node can get stuck in the `waiting_on_user` state indefinitely.
This can occur during `image customization` if any of the following things happen:

- The remote node reboots or crashes.
- The IMS job container on the remote build node is killed or stopped.

## Error identification

A symptom of this problem is a failure when attempting to SSH into the `sshd` container of the IMS image
customization job. Use the following procedure to detect the issue.

> 1. (`ncn-mw#`) Get the details of the image customization job,

Reviewer 3: ```suggestion
1. (`ncn-mw#`) Get the details of the image customization job.
```

  "kubernetes_configmap": "cray-ims-10b93eb5-2926-4521-8cd8-bcb9ab92f989-configmap",
  "kubernetes_job": "cray-ims-10b93eb5-2926-4521-8cd8-bcb9ab92f989-customize",
  "kubernetes_namespace": "ims",
  "kubernetes_pvc": "cray-ims-10b93eb5-2926-4521-8cd8-bcb9ab92f989-job-claim",
  "kubernetes_secret": "cray-ims-10b93eb5-2926-4521-8cd8-bcb9ab92f989-signing-keys",
  "kubernetes_service": "cray-ims-10b93eb5-2926-4521-8cd8-bcb9ab92f989-service",
  "public_key_id": "7d560617-91e1-4075-b8ab-891a3285b783",
  "remote_build_node": "x3000c0s33b1n0",
  "require_dkms": true,
  "resultant_image_id": null,
  "ssh_containers": [
    {
      "connection_info": {
        "cluster.local": {
          "host": "cray-ims-10b93eb5-2926-4521-8cd8-bcb9ab92f989-service.ims.svc.cluster.local",
          "port": 22
        },
        "customer_access": {
          "host": "10b93eb5-2926-4521-8cd8-bcb9ab92f989.ims.cmn.fanta.hpc.amslabs.hpecorp.net",
          "port": 22
        }
      },
      "jail": false,
      "name": "sat-0fda73ef-4ab0-46d2-9e36-9aad6406bb50",
      "status": "pending"
    }
  ],
  "status": "waiting_on_user"
  }
  ```

1. Confirm that the job details fit the symptoms of this issue.

>    - The job `status` field is `waiting_on_user`

Reviewer 3: ```suggestion
   - The job `status` field is `waiting_on_user`.
```

          "port": 22
        },
        "customer_access": {
          "host": "10b93eb5-2926-4521-8cd8-bcb9ab92f989.ims.cmn.fanta.hpc.amslabs.hpecorp.net",
          "port": 22
        }
      },
      "jail": false,
      "name": "sat-0fda73ef-4ab0-46d2-9e36-9aad6406bb50",
      "status": "pending"
    }
  ],
  "status": "waiting_on_user"
  }
  ```

1. Confirm that the job details fit the symptoms of this issue.

   - The job `status` field is `waiting_on_user`
   - The `remote_build_node` field is set, indicating the job is running on a remote build node.

   If either of these is not the case, then the procedure documented here is not applicable.

1. (`ncn-mw#`) Attempt to SSH into the `sshd` container of the job.

   > Perform the following substitutions in the command:
   >
   > - Replace `<IMS_SSH_HOST>` with the value of the
   >    `ssh_containers[0].connection_info.customer_access.host` field in the job details.
   > - Replace `<IMS_SSH_PORT>` with the value of the
   >   `ssh_containers[0].connection_info.customer_access.port` field in the job details.

   ```bash
   ssh -p <IMS_SSH_PORT> root@<IMS_SSH_HOST>
   ```

>   Note: Ensure the user has the correct `ssh` private key for the IMS job.

Reviewer: This should probably be stated before the command, not after it. Do we want to provide any additional guidance beyond this? Refer to some other place in the docs where the ssh key for jobs is discussed (if such a place exists)?

Reviewer 2: Move to before - sure.

The ssh key usage is discussed in sections that talk about creating a job. They need to understand that before the job can be created. It isn't really discussed separately. I think the reminder here should be sufficient.

>    > Perform the following substitutions in the command:
>    >
>    > - Replace `<IMS_SSH_HOST>` with the value of the
>    >    `ssh_containers[0].connection_info.customer_access.host` field in the job details.
>    > - Replace `<IMS_SSH_PORT>` with the value of the
>    >   `ssh_containers[0].connection_info.customer_access.port` field in the job details.
> 
>    ```bash
>    ssh -p <IMS_SSH_PORT> root@<IMS_SSH_HOST>
>    ```
> 
>   Note: Ensure the user has the correct `ssh` private key for the IMS job.

Reviewer: ````suggestion
   > - Ensure that the user running this command has the SSH private key that
   >   is associated with the IMS SSH public key in the `public_key_id` field of the job details.
   > - Perform the following substitutions in the command:
   >     - Replace `<IMS_SSH_HOST>` with the value of the
   >       `ssh_containers[0].connection_info.customer_access.host` field in the job details.
   >     - Replace `<IMS_SSH_PORT>` with the value of the
   >       `ssh_containers[0].connection_info.customer_access.port` field in the job details.

   ```bash
   ssh -p <IMS_SSH_PORT> root@<IMS_SSH_HOST>
   ```
````
